### PR TITLE
Configure Broker to connect to only allowed Historical tiers

### PR DESCRIPTION
**Pull request by kfaraz**
### Description
This PR adds a config for the `CustomTierSelectorStrategy` which can
be used to specify a list of allowed Historical Tiers that the Broker can
connect to.

This can be used to have dedicated Brokers for different data tiers.

### Changes
 * Add `allowedTiers` in `CustomTierSelectorStrategyConfig`
 * Add property `druid.broker.select.tier.custom.allowedTiers`
 * Override server `pick()` methods in `CustomTierSelectorStrategy`
   to only pick servers from allowed tiers
 * Update documentation

<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: docs/configuration/index.md [hunk 1]
```diff
@@ -1605,7 +1605,8 @@ These Broker configurations can be defined in the `broker/runtime.properties` fi
...
+|`druid.broker.select.tier.custom.priorities`|An array of integer priorities. E.g., `[-1, 0, 1, 2]`|Select servers in tiers with a custom priority list.|The config only has effect if `druid.broker.select.tier` is set to `custom`. If `druid.broker.select.tier` is set to `custom` but this config is not specified, the effect is the same as `druid.broker.select.tier` set to `highestPriority`. Any of the integers in this config can be ignored if there's no corresponding tiers with such priorities. Tiers with priorities explicitly specified in this config always have higher priority than those not and those not specified fall back to use `highestPriority` strategy among themselves.|
+|`druid.broker.select.tier.custom.tierNames`|A non-empty array of tier names, e.g., `["_default_tier","_hot_tier"]`|Specifies the tier names to which the Broker can connect. The Broker cannot connect to a Historical process belonging to a tier not listed in this config.|All tiers are allowed if no value is specified for this config or if an empty array is specified. This config has effect only if `druid.broker.select.tier` is set to `custom`.|
```
Comment (abhishekagarwal87) [inline docs/configuration/index.md:1609]:
```suggestion
|`druid.broker.select.tier.custom.allowedTierNames`|A non-empty array of tier names, e.g., `["_default_tier","_hot_tier"]`|Specifies the tier names to which the Broker can connect. The Broker cannot connect to a Historical process belonging to a tier not listed in this config.|All tiers are allowed if no value is specified for this config or if an empty array is specified. This config has effect only if `druid.broker.select.tier` is set to `custom`.|`[]`|
```

File: server/src/main/java/org/apache/druid/client/selector/CustomTierSelectorStrategyConfig.java [hunk 2]
```diff
@@ -31,8 +31,16 @@
   @JsonProperty
   private List<Integer> priorities = new ArrayList<>();
 
+  @JsonProperty
+  private List<String> allowedTiers = new ArrayList<>();
```
Comment (abhishekagarwal87) [inline server/src/main/java/org/apache/druid/client/selector/CustomTierSelectorStrategyConfig.java:35]: this name should be same as what comes after `druid.broker.select.tier.custom.`. so if you are using `druid.broker.select.tier.custom.tierNames`, then this field should be named `tierNames` or you should have the annotation as `@JsonProperty("tierNames")`
Please add a test for serde if not added already.

Comment (kfaraz) [reply server/src/main/java/org/apache/druid/client/selector/CustomTierSelectorStrategyConfig.java:35]: I had changed these names. Must have missed changing here. Thanks!

File: docs/configuration/index.md (hunk 1 above)
Comment (kfaraz) [reply docs/configuration/index.md]: Fixed the name of the config. The last `[]` is not needed as there are already four columns.

**Review by abhishekagarwal87 — APPROVED** [review]

**Comment by dylwylie** [comment]
Is there an advantage to using this new config option over the existing config :: druid.broker.segment.watchedTiers ?

Apprciate that they're not like for like but is there a use case the option added here fulfils that his one doesn't?

**Comment by kfaraz** [comment]
Thanks a lot for the suggestion, @dylwylie . I will try out the config `druid.broker.segment.watchedTiers` and see if this meets all the use cases I had in mind.

**Pull request closed without merging**